Support platform macosx/arm64

allquixotic opened the pull request:
Current version of UnityDoorstop supports only x86 CPU architecture on MacOS. This patch adds support for arm64, which was enabled by integrating the latest version of [plthook upstream](https://github.com/kubo/plthook) into `plthook_osx.c` and re-adding the UnityDoorstop-specific public function, 

```c
void *plthook_handle_by_name(const char *name)
```

... which appears to work just fine after pulling in the changes from plthook upstream.

The build system has been updated to produce a MacOS Universal Binary, by first compiling the x86_64 and arm64 versions of the library, then smashing them together into a single file. You can distribute the Universal Binary as the only MacOS distributable for both Intel and Apple Silicon architectures, and it will automatically "do the right thing" when the user tries to use it, irrespective of their CPU architecture.

NOTE: due to [MonoMod#90](https://github.com/MonoMod/MonoMod/issues/90), game mod loaders that leverage UnityDoorstop as their entry point but use HarmonyX as their managed code patching library will still fail to work on Apple Silicon. This is an issue unrelated to UnityDoorstop, which by all accounts, appears to correctly load and hand over control to the mod loader.

Tested with BepInEx; confirmed that BepInEx's managed code starts running successfully, and it only crashes because of the MonoMod#90 issue mentioned above, which again isn't UnityDoorstop's fault.

Resolving this issue is a **prerequisite** to one day getting mod loaders to work with Apple Silicon on MacOS.

Resolves bug #61.

ManlyMarco reviewed (commented):
On a glance it looks fine but it would be nicer to read without the unnecessary formatting changes, not a big deal though.
The additional logging for config should be a separate PR though.

mikatammi commented:
Are you going to continue working on this? I'd like to see this merged as it eats so much fps if you have to play Valheim on macOS using Rosetta2 x86_64 emulation

ManlyMarco reviewed (changes requested):
Since Harmony and HarmonyX now support ARM, this PR is a lot more relevant.

The changes look good, but linux and macos builds are failing on CI. Once builds are fixed and can be tested, this can be merged.

allquixotic commented:
This is Attempt 1 at fixing the CI build.

## Issues Fixed

### 1. Linux Build Error
**Problem**: Linux builds were failing because they attempted to compile `plthook_osx.c` which includes macOS-specific headers (`mach-o/dyld.h`).

**Solution**: Modified `xmake.lua` to conditionally compile platform-specific plthook files:
- Linux builds now only compile `plthook_elf.c`
- macOS builds now only compile `plthook_osx.c`

### 2. macOS Build Error  
**Problem**: The macOS CI build was failing with "cp: build/macosx/universal/release/.doorstop_version: No such file or directory"

**Solution**: Added proper `.doorstop_version` file generation for all macOS build targets:
- Added `after_build` function to `doorstop_x86_64` target
- Enhanced `doorstop_arm64` target's `after_build` function to:
  - Create its own `.doorstop_version` file
  - Copy the version file to the universal binary directory
  - Ensure the universal directory exists before creating files

## Changes Made to xmake.lua

1. Modified platform detection to only include appropriate plthook files:
   ```lua
   if is_os("linux") then
       add_files("src/nix/plthook/plthook_elf.c")
   elseif is_os("macosx") then
       add_files("src/nix/plthook/plthook_osx.c")
   end
   ```

2. Added `after_build` function to `doorstop_x86_64` target to create version file

3. Replaced wildcard with explicit `plthook_osx.c` for macOS targets

4. Enhanced `doorstop_arm64` target's `after_build` function to properly handle universal binary creation and version files

## Verification

The build has been tested locally and successfully on **MacOS only** (I have access to linux/x64 systems but I'll let CI test that):
- Verified the MacOS build  created all required binaries for both architectures (x86_64 and arm64)
- Manually confirmed it generates `.doorstop_version` files in all expected locations:
  - `build/macosx/arm64/release/.doorstop_version`
  - `build/macosx/x86_64/release/.doorstop_version`
  - `build/macosx/universal/release/.doorstop_version`
  - `build/macosx/universal/debug/.doorstop_version` (for debug builds)
- It created proper universal binaries containing both architectures
- All file copy operations that GitHub Actions performs work correctly

ManlyMarco commented:
Looks good, CI builds are now available for testing at https://github.com/NeighTools/UnityDoorstop/actions/runs/17422686388?pr=62

I'll merge after a few people get a chance to test it on different platforms and confirm everything is working.

arrowmaster commented:
I tested `doorstop_linux_release` on Valheim and it launched with no noticeable issues.

buftio commented:
Joining here! I've tested `universal` build with this PR on The Bazaar (mono .netstandart2.1 build)

and it works fine with BepInEx (manually built for arm) hello world plugin.

Thanks @allquixotic 

Spec: M3 Pro chip
